### ISSUE TITLE
is it possible to mount mov/mp4 as a folder with pngs/jpgs?

### ISSUE THREAD

**Issue by zhuker** [nschlia/ffmpegfs#26]
what would it take to implement it?

like so:
`# ls /storage/videos`

```
video1.mp4
video2.mov
```

`# ffmpegfs /storage/videos /mnt/ffmpegfs`
`# find /mnt/ffmpegfs`

```
/mnt/ffmpegfs/video1.mp4/00001.png
/mnt/ffmpegfs/video1.mp4/00002.png
...
/mnt/ffmpegfs/video1.mov/00001.png
/mnt/ffmpegfs/video1.mov/00002.png
```

**Comment by nschlia** [comment]
Do you mean one image per frame? That'll be thousands of virtual files. This may be possible although I am not sure if FUSE will support that (there is a size limit to the buffer filled in with the files), what's the idea behind that?  

For a 120 second file you could get some 3500 images.  I suppose that could be more than FUSE supports.

**Comment by zhuker** [comment]
we have a video processing pipeline which accepts a bunch of *png/*jpg files
what we do now:
1. decode input video to pngs (store  100 thousand pngs in a folder) - 
2. process video via our neural nets

we waste multiple GBs of storage (pngs need to stay on storage after processing to be able to get back to them)

what i want to achieve
1. mount video with ffmpegfs, get a folder with virtual pngs (no storage wasted)
2. process as usual

our videos are mostly i-frame only (e.g. prores) so it shouldn't be too processor intensive

**Comment by nschlia** [comment]
Now I understand. There are two problems, but maybe we can find a solution:

First it won't save you any disk space. MP4/MOV must be decoded sequentially, so if you want to access say frame number 1234 it would require to decode frame 1 to 1233 first. Of course it is possible to seek to an arbitrary position, at least approximately, and start decoding from there. That's required anyway to have at least one i-frame to generate a full picture. But I am not sure if I could locate that image number 1234 that way. There would be some inaccuracy. So mostly it would mean that the whole file needs to be decoded and stored, either on disk or in memory, waiting for someone to access frame images.

Second problem is, as we do not have a single file that grows, how can you make sure that your neuronal net does not hiccup when it tries to access pngs that are not yet ready? Currently, if an app tries to access a part of the file that has not been decoded the read blocks until that chunk is ready. This could be done with file opens as well, but that means that the accessing software may have to wait even for minutes to be able to get the image, depending on the size of the video...

A solution for problem 1 could be to disable the cache but that means once an image was read it will be discarded. Accessing it again means decoding it again. That's a processing time hog.

Problem 2 depends on your software. If it tolerates that it would work.

So some questions:

Will your software read the pngs sequentially and only once? If so, it could be done. Images must still be cached but only until read and then could be discarded. 

If it tries to read faster than the images can be made available, will it accept read blocks, even if for several seconds? This will happen especially if it skips images, for example, only reads an image every 5 seconds or so.

Possible trouble:

I'll have to try how FUSE handles 1000+ virtual files. Maybe it cannot handle that and then it'll be a no-go :(

Suggestion:

The cache and discard after reading function would require some significant changes, but I could implement the video-to-image thing with normal caching first so you could try out if it works for you at all.

**Comment by zhuker** [comment]
first of all - thanks for giving it a thought and for your time

> First it won't save you any disk space. MP4/MOV must be decoded sequentially, so if you want to access say frame number 1234 it would require to decode frame 1 to 1233 first.

In general case - you are right. But that's not really a problem for us, since we are mostly using I-frame only video formats (e.g. ProRes+mov and JPEG2000+mxf) which are easily seekable.

> But I am not sure if I could locate that image number 1234 that way.

again - in general case you are right. but for our combination of codec+container frames are quite easy to locate

>  Second problem is, as we do not have a single file that grows, how can you make sure that your neuronal net does not hiccup when it tries to access pngs that are not yet ready?

That's fine, it'll block and wait until file is ready

> wait even for minutes to be able to get the image, depending on the size of the video...

waiting for _minutes_ isn't great but it's way better than the several hours we wait before we start processing today

>  A solution for problem 1 could be to disable the cache but that means once an image was read it will be discarded. Accessing it again means decoding it again. That's a processing time hog.

our videos are normally 4K HDR videos (16bit per pixel) so 20MB per _frame_ file size is not rare, i would very much like to keep caching even if it means storing some of the pngs on actual disk (some LRU policy?)  
 
> Will your software read the pngs sequentially and only once? 

the processing pipeline is sequential, but when it fails we want to analyze why something happened and then out of this 100k images we want to randomly take a look at 100-200 hundred. Some kind of cache is still needed, me thinks.

> If it tries to read faster than the images can be made available, will it accept read blocks, even if for several seconds? This will happen especially if it skips images, for example, only reads an image every 5 seconds or so.

we do process every single frame (no skips), but blocking reads are totally cool (even for seconds), we dont need to be realtime or anything like that

> I'll have to try how FUSE handles 1000+ virtual files. Maybe it cannot handle that and then it'll be a no-go :(

please let me know.

i have no idea how FUSE works internally but shouldn't it be
1. multiple `readdir`-s to list entries 
2. `open` or `stat` specific file 

let's assume the following:
- we only support easily seekable formats. on mount `ffmpegfs` will check if video is of supported type and refuse to mount otherwise
- fs will be mounted read-only 

for `readdir` you could keep some kind of simple cache 

for `open` you quickly seek (because we only support seekable video) and decode file

i have a problem understanding how `stat` would work. but for simplicity sake let's say images are going to be `.bmp` (or any other uncompressed format) not `.png` so we know ahead of time what the file size would be 

can you educate me why given these assumptions you would need to take care of 1000+ fuse virtual files?

> The cache and discard after reading function would require some significant changes, but I could implement the video-to-image thing with normal caching first so you could try out if it works for you at all.
  
that would be AMAZING

**Comment by nschlia** [comment]
> let's assume the following:
> 
> - we only support easily seekable formats. on mount ffmpegfs will check if video is of supported type and refuse to mount otherwise
> - fs will be mounted read-only
> 
> for readdir you could keep some kind of simple cache
> 
> for open you quickly seek (because we only support seekable video) and decode file
> 
> i have a problem understanding how stat would work. but for simplicity sake let's say images are going to be .bmp (or any other uncompressed format) not .png so we know ahead of time what the file size would be

Given all these assumptions it should be feasible. For stat there may be a problem with formats like png/jpg, because I can not provide a predicted size until the image is actually ready. So the size will definitely be wrong until the image is available, unless I use a fixed size format like BMP which will be a major waste of disk space :) 

Your software that consumes the files must cope with files that change size after opening. They may be smaller or larger. 

This is probably a big issue when you access the files via Samba or NFS.  Samba fills files smaller than predicted with zeros to this size, or cuts them to the size (discards the end) if it is larger. NFS also fills smaller files with zeroes, larger files will either be cut or correct (this depends, you can try it as many times as you want to, the outcome is not deterministic).

This is why I add 2.5% to the size prediction to raise the possibility of the file ending up a bit smaller to avoid the cut off... So we could set this to 5 or more per cent to avoid list parts. On the other hand this would only mean that a very small portion at the bottom of the image is missing and probably nothing important. 

> can you educate me why given these assumptions you would need to take care of 1000+ fuse virtual files?

Once someone does a ls on a virtual directory, FUSE issues a readdir to me and I have to present the whole directory. So for a 10.000 frames video I would have to present 10.000 virtual images. There seem to be ways to do that with FUSE but I need to try out what happens...

PS:

I am currently preparing release 1.7 (and actually 1.8 which only adds Doxygen documentation), therefore I currently have a feature freeze. I cannot add new features at the moment but I would create a push request, implement your features and then merge it back later. The functionality sounds interesting, I would like to have it in FFmpegfs. I also don't know how long it will take to realise it, but probably a week or two.

**Comment by zhuker** [comment]
> Given all these assumptions it should be feasible. For stat there may be a problem with formats like png/jpg, because I can not provide a predicted size until the image is actually ready. So the size will definitely be wrong until the image is available, unless I use a fixed size format like BMP which will be a major waste of disk space :)

> format like BMP which will be a major waste of disk space :)

you mean for cache? i am not copying the files anywhere, just reading them from `ffmpegfs`

>  Your software that consumes the files must cope with files that change size after opening. They may be smaller or larger.

i vote for BMP or any other uncompressed format (e.g. PGM, PNM, PPM, uncompressed TIF) so we know the size ahead of time. or at least there should be an option (compressed/uncompressed)

> we could set this to 5 or more per cent to avoid lost parts

thats cool

> very small portion at the bottom of the image is missing and probably nothing important.

neural nets normally give unpredictable results  given damaged/incomplete input, so it's important to preserve the entire image

> FUSE issues a readdir to me and I have to present the whole directory

i see! 
there's no incremental `readdir` in FUSE?

> The functionality sounds interesting, I would like to have it in FFmpegfs.

Yeah that would be really cool. 

> I also don't know how long it will take to realise it, but probably a week or two.

I wish I could help you, but my knowledge of C++ is really bad. 
I'll be able to fix minor bugs and make some tweaks.

**Comment by nschlia** [comment]
> > format like BMP which will be a major waste of disk space :)
> 
> you mean for cache? i am not copying the files anywhere, just reading them from `ffmpegfs`
> 
For cache. FFmpegs needs to store the images somewhere, memory or disk, at least for as long as they have to be available. Redecoding them over every time someone wants to read them is far too time consuming.

> i vote for BMP or any other uncompressed format (e.g. PGM, PNM, PPM, uncompressed TIF) so we know the size ahead of time. or at least there should be an option (compressed/uncompressed)

I'll create PGM, PNM, BMP, or whatever you prefer, this is a simple codec setting, ffmpeg API does that for me. Just choose your favourite weapon :)

> > very small portion at the bottom of the image is missing and probably nothing important.
> 
> neural nets normally give unpredictable results given damaged/incomplete input, so it's important to preserve the entire image

OK, the we need to go for a predictable format. We can then try PNG later or whatever and see what happens.

> there's no incremental `readdir` in FUSE?

No, that's the design of the file systems that access the FUSE drive. You are required to provide the complete directory. Even if you do "ls notEXISTING" the command still needs to examine all files.

**Comment by zhuker** [comment]
sounds good

**Comment by nschlia** [comment]
Created a [branch](https://github.com/nschlia/ffmpegfs/tree/FB_issue_%2326) to handle this.

**Comment by zhuker** [comment]
just found out `-compression_level 0 -f image2 '%06d.png'` gives predictable size pngs but of course uncompressed

**Comment by nschlia** [comment]
Work in progress...

**Comment by nschlia** [comment]
The desired functionality has been implemented now, at least partly. There are a few things missing but the current state should be sufficient as a proof of concept.

**Caution: The code is by no means yet production grade, it needs extensive overhaul yet, but it works.**

The functionality that is missing or incomplete:

1. It is currently not possible to access a file directly without doing a "ls" on the parent directory first.
2. The cache format is a ludicrous waste of disk space and needs to be improved.
3. The cache format change should be automatically applied, for now it is required to delete the cache directory manually before using this version.
4. The source must be all p-frame, incomplete images will be stored they way they are.

For 1.: A simple
```
ls /mnt/ffmpegfs/video1.mp4/00001.png
```
will fail. You need to list the parent directory first like:
```
ffmpegfs /storage/videos /mnt/ffmpegfs
find /mnt/ffmpegfs

/mnt/ffmpegfs/video1.mp4/00001.png
/mnt/ffmpegfs/video1.mp4/00002.png
...
/mnt/ffmpegfs/video1.mov/00001.png
/mnt/ffmpegfs/video1.mov/00002.png
```
This builds the frame image virtual directories. Once this was done, the "ls" command will succeed. I guess that won't be a problem as the software that consumes the frame images needs to list the directories in the first place anyway to find out what to scan...

For 2.: For each image a 2 MB segment is reserved in cache. That makes it easy to access e.g. image number 1234 - it is simply at 1234 * 2 MB. But that's a big waste of disk space and limits image size to 2 MB.
For 3.: Do a "rm -Rf /path/to/cache" before running ffmpegfs.
For 4.: i-frames can be completed with information of previous frames. There are filters to do that, they just need to be activated inside the code.

If the functionality proves useful, these ristrictions will be removed later.

**Special considerations:**

Creating a virtual set of images won't save you any disk space. The frame images have to be stored somewhere. So you need to set the cache path to a drive with enough space (--cachepath=DIR). Also set a cache limit that is sufficent or remove it completely (--max_cache_size=0). You may set cache expiry time to 1 day (--expiry_time=1d) or whatever appropriate for you so old files get cleaned up once no longer required.

The biggest advantage of using a virtual file set is the ease of use - no need to convert the video to frame images first. Just access them when required and they can be used as soon as they are ready.

To get the right code, change the branch to FB_issue_#26 and download the ZIP or if you "git clone" it don't forget to do "git checkout FB_issue_#26" like I did once... :)

**Comment by zhuker** [comment]
Thanks a lot! I will give a try now.
Quick question: why 2MB limit? our frames are normally 1920x1080 which is already over 2MB uncompressed

**Comment by zhuker** [comment]
Doesn't mount my folder with videos 
see my shell history attached
<img width="1304" alt="Screen Shot 2019-03-20 at 2 52 34 PM" src="https://user-images.githubusercontent.com/4367682/54685589-d7512700-4b1f-11e9-8d19-667243808621.png">

**Comment by zhuker** [comment]
I am on `FB_issue_#26` branch

```
zhukov@kote:~/ffmpegfs$ git branch
* FB_issue_#26
master
```

**Comment by nschlia** [comment]
> 
> 
> Thanks a lot! I will give a try now.
> Quick question: why 2MB limit? our frames are normally 1920x1080 which is already over 2MB uncompressed
If you use JPG or PNG it should work. If this is not enough you can change fileio.h line 60:
`#define IMAGE_MAX_SIZE          (2*1024*1024)`
You could set 5 or more MB. But I guess for PNG/JPG that should be sufficient.

**Comment by zhuker** [comment]
> > Thanks a lot! I will give a try now.
> > Quick question: why 2MB limit? our frames are normally 1920x1080 which is already over 2MB uncompressed
> > If you use JPG or PNG it should work. If this is not enough you can change fileio.h line 60:
> > `#define IMAGE_MAX_SIZE          (2*1024*1024)`
> > You could set 5 or more MB. But I guess for PNG/JPG that should be sufficient.

Our frames converted to PNG are normally 10-20MB each

**Comment by nschlia** [comment]
> Doesn't mount my folder with videos
> see my shell history attached

Did you set --desttype=png or --desttype=jpg? If you set the target to anything else than png/jpg/bmp, you will get audio and video files.


--cachepath=/path/to/huge/drive --max_cache_size=0 --expiry_time=1d

**Comment by zhuker** [comment]
oh my god!!! it works!!!

**Comment by zhuker** [comment]
Too early to celebrate :(
it does list frames correctly
but here is what i get if i want to `ffmpeg -i` on a PNG
 
<img width="702" alt="Screen Shot 2019-03-20 at 3 08 19 PM" src="https://user-images.githubusercontent.com/4367682/54686467-0a94b580-4b22-11e9-8c86-d343b86eb1f1.png">
<img width="879" alt="Screen Shot 2019-03-20 at 3 08 27 PM" src="https://user-images.githubusercontent.com/4367682/54686473-0d8fa600-4b22-11e9-8206-316858415aca.png">

**Comment by nschlia** [comment]
> 
> 
> > > Thanks a lot! I will give a try now.
> > > Quick question: why 2MB limit? our frames are normally 1920x1080 which is already over 2MB uncompressed
> > > If you use JPG or PNG it should work. If this is not enough you can change fileio.h line 60:
> > > `#define IMAGE_MAX_SIZE          (2*1024*1024)`
> > > You could set 5 or more MB. But I guess for PNG/JPG that should be sufficient.
> 
> Our frames converted to PNG are normally 10-20MB each

Whoopsy! Well in that case you cold set 
`#define IMAGE_MAX_SIZE (25*1024*1024)`
but that would eat up a lot of disk space. I guess with the compression rate I've set 2 MB should be enough. I'll rework the cache format when everything else works.

**Comment by zhuker** [comment]
> `#define IMAGE_MAX_SIZE (25*1024*1024)`
just set to 20MB

**Comment by nschlia** [comment]
> Too early to celebrate :(
> it does list frames correctly
> but here is what i get if i want to `ffmpeg -i` on a PNG

That's what I was afraid of... The pixel format of your source video is not supported by PNG. Maybe JPG works... If not one of the next things I wanted to do is convert the pixel format if required. This is already done for video to video conversion.

Maybe you can use JPG or you can find a video that has a different pixel format.

If not I am sorry you'll have to be patient and wait a few days until I have completed the pixel format conversion.

**Comment by zhuker** [comment]
well `png` is always `rgb24` or `rgb48` and most videos are `yuv420` or `yuv422p10` pixel format conversion should be there

**Comment by nschlia** [comment]
OK. Sorry. Then please be patient I'll add the format conversion.

**Comment by zhuker** [comment]
i am not complaining it's already a miracle :)

**Comment by zhuker** [comment]
`desttype=jpg` works!

**Comment by zhuker** [comment]
but even for i-frame only videos it takes FOREVER to open the last frame

**Comment by nschlia** [comment]
> 
> 
> `desttype=jpg` works!

Great. But the pixel format conversion is still required. I also have many videos that won't work. As a nice little extra I can also apply deinterlacing and complete p-frames. So this is a must have.

> but even for i-frame only videos it takes FOREVER to open the last frame

The code is far from being optimal, especially the caching code is crap. I know. I wanted to prove if this is feasable in the first place, seems it is, so it's worth digging deeper now. There is much room for optimisations. 

Maybe you can check if your video processing software likes the results, especially these locked files until available and the size "morphing". I can speed up processing but not that the files will be there in an instant. Also I will never be able provide the exact file size from the start.

If you video software basically copes with that it's worth walk the extra mile from here.

**Comment by nschlia** [comment]
> 
> 
> i am not complaining it's already a miracle :)

I know :)

**Comment by zhuker** [comment]
here is a sample 10bit `yuv422p10` prores https://kote.videogorillas.com/vmir/vdms/orig.mov
when i mount it with `desttype=jpg` 
it produces this:
![00000000001](https://user-images.githubusercontent.com/4367682/54687775-06b66280-4b25-11e9-9837-8ffc859b734a.jpg)

i assume its because of lack of `yuv422p10` to `yuvj420` conversion

**Comment by zhuker** [comment]
when i mount `yuv420p` h264 https://kote.videogorillas.com/vmir/vdms/orig.mp4
it produces correct jpg
![00000000001](https://user-images.githubusercontent.com/4367682/54687871-4bda9480-4b25-11e9-9d42-a4baa619cf44.jpg)

**Comment by zhuker** [comment]
but  all that aside - ITS AMAZING!!! 🥇 👍 🏆

**Comment by nschlia** [comment]
Is that a scene from "Animal House"?

Well, as I said, I'll implement the pixel conversion. This is going to work with all formats, with a small degradation in quality, but I guess JPG or PNG compression will do more harm.

More to come... :)

**Comment by zhuker** [comment]
its a scene from "Dawson Creek" old time tv show

**Comment by zhuker** [comment]
what are you using (editor/debugger/ide) to write code?

**Comment by nschlia** [comment]
I use QtCreator, with a few tricks I can write code and debug, of course without Qt. I even use QtCreator to write (non-Qt) code for ARM using a cross compiler.

**Comment by zhuker** [comment]
`            make_file(buf, filler, VIRTUALTYPE_FRAME, origpath, filename, 40 * 1024, virtualfile->m_st.st_ctime);`
this is where 40KB size comes from!
why not 42KB then :)

**Comment by nschlia** [comment]
> this is where 40KB size comes from!
> why not 42KB then :)
Well, 42 is the answer to all questions.

You may give it another try. I have implemented pixel format conversion, so your ProRes videos should do as source. Actually you have the full program: pixel format conversion, deinterlacing and rescaling.

Currently jpg works only, png and bmp (RGB pixel formats) yield strange results. It seems I have a lack of understanding how RGB pixel formats work. I try to figure out how.

I also change the max. image size to 20 MB.

There's still a lot to do, but maybe now you can evaluate the result so far.

**Comment by nschlia** [comment]
Found this, that's RGB to YUV though... https://stackoverflow.com/questions/21938674/ffmpeg-rgb-to-yuv-conversion-loses-color-and-scale

**Comment by nschlia** [comment]
More:

https://stackoverflow.com/questions/21938674/ffmpeg-rgb-to-yuv-conversion-loses-color-and-scale?answertab=votes#tab-top

**Comment by nschlia** [comment]
It seems that the deinterlace filter does not work. Simply do not use it, and it will work. I'll try to find out what the problem is. Seems that I am using it wrong. I should work on RGB frames.

**Comment by zhuker** [comment]
Thanks. I’ll give it a try. Not sure I’ll be able to test before Monday. But if I do have a chance, I’ll report back.
Are you using swscale to do pixel format conversions?
Deinterlace is not needed in our processing , we do our own.
https://github.com/A-Bush/Deep-Video-Deinterlacing/blob/master/README.md

**Comment by nschlia** [comment]
Found the problem. Deinterlacing now also works! You can check now if it works for you, next I'll go into optimisations:

- Speed up processing
- Enhance cache format (file size and speed)
- Predict file sizes (currently arbitrarily set to 40 Kb)
- Allow "bitrate" selection. i.e, allow selection of compression factor

**Comment by nschlia** [comment]
> Thanks. I’ll give it a try. Not sure I’ll be able to test before Monday. But if I do have a chance, I’ll report back. Are you using swscale to do pixel format conversions? Deinterlace is not needed in our processing , we do our own. https://github.com/A-Bush/Deep-Video-Deinterlacing/blob/master/README.md

Pixel format conversion is done with sws_scale. If the image size is not limited by command line the function merely does the pix_fmt conversion. 

If you don't need deinterlace, simply do not use the --deinterlace parameter and the filter will not be called at all.

So I'll wait for the results now. If I find some time maybe I'll get into the list of optimisations.

**Comment by nschlia** [comment]
> Deinterlace is not needed in our processing , we do our own. https://github.com/A-Bush/Deep-Video-Deinterlacing/blob/master/README.md

I see. For simplicity I currently only support yadif, but FFmpeg API supports many algorithms,  nnedi also uses a neural network. Or there's one that can use CUDA for speed up.

Maybe one of those could save you one processing step. If you could use on just open an issue for it.

Deinterlace the input video ("bwdif" stands for "Bob Weaver Deinterlacing Filter"). 
https://ffmpeg.org/ffmpeg-filters.html#toc-bwdif
Deinterlace input video by applying Donald Graft’s adaptive kernel deinterling. Work on interlaced parts of a video to produce progressive frames. 
https://ffmpeg.org/ffmpeg-filters.html#toc-kerndeint
Apply motion-compensation deinterlacing. 
https://ffmpeg.org/ffmpeg-filters.html#toc-mcdeint
Deinterlace video using neural network edge directed interpolation. 
https://ffmpeg.org/ffmpeg-filters.html#toc-nnedi
Deinterlace the input video ("w3fdif" stands for "Weston 3 Field Deinterlacing Filter"). 
https://ffmpeg.org/ffmpeg-filters.html#toc-w3fdif
Deinterlace the input video ("yadif" means "yet another deinterlacing filter").
https://ffmpeg.org/ffmpeg-filters.html#toc-yadif-1
Deinterlace the input video using the yadif algorithm, but implemented in CUDA so that it can work as part of a GPU accelerated pipeline with nvdec and/or nvenc. 
https://ffmpeg.org/ffmpeg-filters.html#toc-yadif_005fcuda

**Comment by zhuker** [comment]
ok finally got to test it, sorry it took me so long
it works! 👍 

prores -> png ok but should be rgb48 (not rgb24) because prores has 10bit color 
prores -> jpg ok but weird result when viewed on mac (see below)
h264 -> png ok 
h264 -> jpg same weird result on mac (see below)



prores->png
![00000000001](https://user-images.githubusercontent.com/4367682/55006466-7aa1b080-4fe6-11e9-87a6-8c79259faf13.png)

prores->jpg
![00000000001](https://user-images.githubusercontent.com/4367682/55006471-7d9ca100-4fe6-11e9-9226-3bbdb6808aa6.jpg)

when viewed on Mac
<img width="1807" alt="Screen Shot 2019-03-26 at 4 48 55 PM" src="https://user-images.githubusercontent.com/4367682/55006848-16cbb780-4fe7-11e9-898d-3d9c195ccb85.png">

h264(yuv420p)->png
![00000000002](https://user-images.githubusercontent.com/4367682/55006670-c9e7e100-4fe6-11e9-972c-3ddc0866a6c7.png)
h264(yuv420p)->jpg
![00000000002](https://user-images.githubusercontent.com/4367682/55006715-dc621a80-4fe6-11e9-8505-da6882ae0c94.jpg)

when viewed on Mac
<img width="1794" alt="Screen Shot 2019-03-26 at 4 50 37 PM" src="https://user-images.githubusercontent.com/4367682/55006982-4da1cd80-4fe7-11e9-89cc-e05d88508f19.png">

**Comment by nschlia** [comment]
> ok finally got to test it, sorry it took me so long
> it works! +1
> 
Hooray :)

> prores -> png ok but should be rgb48 (not rgb24) because prores has 10bit color
FFmpeg only seems to have rgb48be (rgb48 big endian format). I could give it a try. I'll implement a selection of the best match format (using av_find_best_pix_fmt_of_2 or so). But that's not on top of the list, there are a few other things more important to be completed first.

> prores -> jpg ok but weird result when viewed on mac (see below)
> h264 -> png ok
> h264 -> jpg same weird result on mac (see below)

It should not make a difference which source file format you use (precisely which codec, H264, ProRes or whatever) as the pixel format is converted to what PNG, JPG or BMP requires. So it's no wonder the weird results come up with every source.

Have you opened the images directly from the ffmpegfs virtual directory? What happens if you local copy the images on your mac and then open that copy? What happens if you open the virtual image a second time? Still distorted? If so the mac viewer does not like the images I create...  

What about BMP?

On Ubuntu the stock image viewer (Gnome Desktop) showed only a small stripe of the image top until I added the functionality that updates the file size once it is known. You can see that when you open an image in KDE Dolphin or Windows Explorer, the file size gets refreshed (from 40 KB to 6 MB or so). In Dolphin you may have to refresh (F5), Explorer does that automatically. 

Anyway, the only tweak compared to a real image in a real directory is the file size: ffmpegfs pretends all images to be 40 KB until it has been opened (e.g. with an image viewer) and decoded. Then the file size magically changes to the real size.

Anyway, there is still a lot to do. I am glad that ffmpegfs basically works for you. The list so far:

- Allow selecting the target image quality (using --videobitrate= or a new parameter)
- Predict the target image size at least approximately (currently fixed to an arbitrary 40 K)
- Update the file sizes once known
- Change the internal cache format to support any size of images (remove the 20 M limit)
- Speed up caching
- Select the best match target pixel format (use av_find_best_pix_fmt_of_2 or so)

I will look into that this weekend.

**Comment by zhuker** [comment]
when viewing on mac i copied from the virtual directory first by just using `cp 000000001.png ~`

**Comment by zhuker** [comment]
one more thing i am missing is ability to read random frames without waiting for decode of previous ones. this would be very handy for debugging.
also `find /mnt/ffmpegfs` is not feasible on our 100TB video storage 
would be nice to just be able to:
`cp /mnt/ffmpegfs/long/path/to/video/here/00000001.png ~`

**Comment by zhuker** [comment]
> Predict the target image size at least approximately (currently fixed to an arbitrary 40 K)

not sure this will work with jpegs 
but for png if i could specify some option like `--exact-sizes` it would turn on ffmpeg option '-compression_level 0` and the size would be 100% predictable `header + width * height * bytes_per_pixel`

**Comment by nschlia** [comment]
> one more thing i am missing is ability to read random frames without waiting for decode of previous ones. this would be very handy for debugging.

I guess this should be possible. For video or audio files it's almost impossible but as we want separate, independent frame images it could be done. Added to the list.

> also `find /mnt/ffmpegfs` is not feasible on our 100TB video storage
> would be nice to just be able to:
> `cp /mnt/ffmpegfs/long/path/to/video/here/00000001.png ~`

This is on my list, I forgot to mention.

>not sure this will work with jpegs
>but for png if i could specify some option like --exact-sizes it would turn on ffmpeg option >-compression_level 0 and the size would be 100% predictable header + width * height * bytes_per_pixel

Added.

This is the complete list now: 

I am afraid this is going to take me a few weeks to complete, so please stay tuned :)

- Allow direct access without prio ls (`cp /mnt/ffmpegfs/long/path/to/video/here/00000001.png ~`)
- Allow selecting the target image quality (using --videobitrate= or a new parameter)
- Predict the target image size at least approximately (currently fixed to an arbitrary 40 K)
- For png if it could specify some option like --exact-sizes it would turn on ffmpeg option -compression_level 0 and the size would be 100% predictable header + width * height * bytes_per_pixel
- Update the file sizes once known
- Change the internal cache format to support any size of images (remove the 20 M limit)
- Speed up caching
- Select the best match target pixel format (use av_find_best_pix_fmt_of_2 or so)
- Ability to read random frames without waiting for decode of previous ones

**Comment by nschlia** [comment]
> when viewing on mac i copied from the virtual directory first by just using `cp 000000001.png ~`

And this as well. Seems your Mac does not like my image format. Have to figure out why.

**Comment by nschlia** [comment]
> > when viewing on mac i copied from the virtual directory first by just using `cp 000000001.png ~`
> 
> And this as well. Seems your Mac does not like my image format. Have to figure out why.

I tried that on my (oldie Snow Leopard OS) and the images were OK. Did you do the cp twice? Maybe the first time it copy 40 KB only. Still have to fix the size...

Maybe you want to wait a few days, I just fixed the pixel format conversion. Now a best match format is chosen, for PNG it selects RGB48BE now when I convert from ProRes, for example.

**Comment by zhuker** [comment]
re images on mac - the very same bit by bit png file opens ok in Gimp but not ok in Preview
i assume it's the declared and actual pixel formats

**Comment by nschlia** [comment]
> re images on mac - the very same bit by bit png file opens ok in Gimp but not ok in Preview
> i assume it's the declared and actual pixel formats

Maybe the next update fixes it. I also added the direct access thing so you don't need to ls first.
Unfortunately I'll be in Chicago for the next one or two weeks and I am not sure if I am able to work on this.

You can send me an email to nschlia@oblivion-software.de if you want to so I can keep you up to date without bloating this issue.

**Comment by nschlia** [comment]
Here's the update with the following new functionality for video frame images:

* Random access to any image
* Improved cache supporting images of any size

The random access strategy works as follows:

When a virtual directory with frame images is opened, ffmpegfs starts decoding from the beginning of the file. If any image of the list is opened, it skips to 25 frames before it and starts decoding, (almost) immediately providing the requested image. If the defaults are not changed (which is recommended), decoding is suspended after 30 seconds and terminated after 60 seconds of inactivity. Once there is another frame image accessed, decoding continues or restarts those 25 frames before that very image. 

This should save a lot of disk space as only the images requested get decoded plus an overhead of maybe 100 to 150 frames. That means that normally the whole file will never be recoded, only a fraction of it unless everything is accessed. We might be talking of hundreds of megabytes, mayb even giga- or terabytes that would fill up.

I decided to start 25 frames earlier as it only takes a fraction of a second and maybe the user wants to go back a few frames. If not, we are talking about a few hundred milliseconds wasted, so who should care.

Notes:

* I refrained from creating uncompressed PNGs. The format is lossless, encoding time seems to be negligible, so better save valuable disk space.
* The first image is repeated once, so if frame no. 200 is accessed it is actually 199 but I guess this does not really matter.

To create test files, run these commands (see test/mkvid):

```
ffmpeg -y -f lavfi -i color=c=blue:s=1920x1280:d=60:r=30000/1001 -g 1 -c:v prores -vf \
 "drawtext=fontsize=800: \
  fontcolor=white:x=(w-text_w)/2:y=(h-text_h)/2: \
  text='%{frame_num}': start_number=1" \
 frame_test_ntsc.mov

ffmpeg -y -f lavfi -i color=c=blue:s=1920x1280:d=60:r=25 -g 1 -c:v prores -vf \
 "drawtext=fontsize=800: \
  fontcolor=white:x=(w-text_w)/2:y=(h-text_h)/2: \
  text='%{frame_num}': start_number=1" \
 frame_test_pal.mov
```

 This will create 60 second movie at 1920x1280 resolution, where each frame contains its number. The NTSC version at ~29.997 frames per second ist a good test to check if the frame to pts calculations and vice versa work on non-integer frame rates.

**Comment by nschlia** [comment]
There are still some things to do:

* Speed is still an issue. For some reason decoding is quite slow, I'll have to figure out why.
* Need to do some testing, all works fine under Debian KDE and Ubuntu Gnome. Tests with Windows and Apple OS must still be made.
* There's much room for optimisations.

BTW, I like that random access thing. I've been missing that for video and audio as well. I'll extend that feature soon. Using the enhanced caching code this should be feasible.

**Comment by nschlia** [comment]
* Enhanced caching strategy: Already decoded frames will not be recoded, saving a lot of time. Frames that already exist in cache are simply skipped until EOF or missing frames are found. 
* The prescan cache was reduced to 3 frames as with the new strategy access is very fast.
* Enabled profile guided optimisation. This is a big trade of: Compile time to runtime. Building takes much longer now, but the resulting code should be optimised to the machine its built on.

**Comment by nschlia** [comment]
Currently all that's left is the size prediction and that the first frame is duplicated, so each image frame is one behind actually. But this should both be only minor flaws.

**Comment by nschlia** [comment]
Found the reason for the duplicated first frame: Happens when deinterlacing is on. If it's turned off, the frames are correct. So it's not a bug, it's a feature.

**Comment by nschlia** [comment]
Found two problems:

* Sometimes the real number of frames falls a bit shorter than the predicted value. If e.g. the source file contains e,g. 30 frames, but 32 were predicted, the client sometimes waits forever if it accesses frame 31 or 32. I'll have to find a way to correct this number once known.
* The cache works well on most source files but I found at least one where it goes berserk. Although it seeks to the right position (if already decoded frames), it only reports errors and only returns some of the results.

I'll have to investigate that and find a fix.

**Comment by zhuker** [comment]
locks up at latest commit 4328b66bc53c3c479ae373c4ac28ad83354258fa

1. mount a folder with https://kote.videogorillas.com/vmir/uk/demo/R3A-shot54-sbs.mp4 `ffmpegfs -d --cachepath=/home/zhukov/cacheffmpegfs --expiry_time=1d  --desttype=png /storage/vmir/uk/demo/ /home/zhukov/mnt/`
2. `ls mnt/R3A-shot54-sbs.mp4` - ok 💚 
3.  `cp mnt/R3A-shot54-sbs.mp4/00000000067.png ~` locks up 🛑 

```
open flags: 0x8000 /R3A-shot54-sbs.mp4/00000000067.png
2019-05-06 05:31:23 DEBUG  : [/storage/vmir/uk/demo/R3A-shot54-sbs.mp4] Starting decoder thread.
2019-05-06 05:31:23 INFO   : [/storage/vmir/uk/demo/R3A-shot54-sbs.mp4] Transcoding to png. Predicted size unlimited (0 bytes).
2019-05-06 05:31:23 INFO   : [/storage/vmir/uk/demo/R3A-shot54-sbs.mp4] Opening input file.
2019-05-06 05:31:23 DEBUG  : [/storage/vmir/uk/demo/R3A-shot54-sbs.mp4] Opened input codec for stream #0: H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
2019-05-06 05:31:23 INFO   : [/storage/vmir/uk/demo/R3A-shot54-sbs.mp4] Video in: h264@37.02 Mbps [00:02.760]
2019-05-06 05:31:23 WARNING: [/storage/vmir/uk/demo/R3A-shot54-sbs.mp4] Unsupported video codec 'png' for format png.
2019-05-06 05:31:23 INFO   : [/home/zhukov/mnt/R3A-shot54-sbs.png] Opening output file.
2019-05-06 05:31:23 DEBUG  : [/home/zhukov/mnt/R3A-shot54-sbs.png] Opening format type 'png'.
2019-05-06 05:31:23 DEBUG  : [/home/zhukov/mnt/R3A-shot54-sbs.png] Output pixel format: rgb24
2019-05-06 05:31:23 DEBUG  : [/home/zhukov/mnt/R3A-shot54-sbs.png] Pre-buffering up to 102400 bytes.
[h264 @ 0x7fb08c003100] illegal short term buffer state detected
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] illegal short term buffer state detected
2019-05-06 05:31:23 DEBUG  : [/home/zhukov/mnt/R3A-shot54-sbs.png] Pre-buffer limit reached.
2019-05-06 05:31:23 DEBUG  : [/storage/vmir/uk/demo/R3A-shot54-sbs.mp4] Decoder thread is running.
   open[140396172708752] flags: 0x8000 /R3A-shot54-sbs.mp4/00000000067.png
   unique: 91, success, outsize: 32
unique: 92, opcode: READ (15), nodeid: 75, insize: 80, pid: 26316
read[140396172708752] 131072 bytes from 0 flags: 0x8000
[h264 @ 0x7fb08c003100] illegal short term buffer state detected
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] illegal short term buffer state detected
[h264 @ 0x7fb08c003100] reference picture missing during reorder
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] reference picture missing during reorder
[h264 @ 0x7fb08c003100] Missing reference picture, default is 22
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] Missing reference picture, default is 22
[h264 @ 0x7fb08c003100] illegal short term buffer state detected
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] illegal short term buffer state detected
[h264 @ 0x7fb08c003100] reference picture missing during reorder
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] reference picture missing during reorder
[h264 @ 0x7fb08c003100] Missing reference picture, default is 22
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] Missing reference picture, default is 22
[h264 @ 0x7fb08c003100] reference picture missing during reorder
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] reference picture missing during reorder
[h264 @ 0x7fb08c003100] Missing reference picture, default is 24
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] Missing reference picture, default is 24
[h264 @ 0x7fb08c003100] illegal short term buffer state detected
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] illegal short term buffer state detected
[h264 @ 0x7fb08c003100] reference picture missing during reorder
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] reference picture missing during reorder
[h264 @ 0x7fb08c003100] Missing reference picture, default is 24
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] Missing reference picture, default is 24
[h264 @ 0x7fb08c003100] reference picture missing during reorder
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] reference picture missing during reorder
[h264 @ 0x7fb08c003100] Missing reference picture, default is 26
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] Missing reference picture, default is 26
[h264 @ 0x7fb08c003100] illegal short term buffer state detected
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] illegal short term buffer state detected
[h264 @ 0x7fb08c003100] reference picture missing during reorder
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] reference picture missing during reorder
[h264 @ 0x7fb08c003100] Missing reference picture, default is 26
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] Missing reference picture, default is 26
[h264 @ 0x7fb08c003100] reference picture missing during reorder
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] reference picture missing during reorder
[h264 @ 0x7fb08c003100] Missing reference picture, default is 28
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] Missing reference picture, default is 28
[h264 @ 0x7fb08c003100] illegal short term buffer state detected
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] illegal short term buffer state detected
[h264 @ 0x7fb08c003100] reference picture missing during reorder
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] reference picture missing during reorder
[h264 @ 0x7fb08c003100] Missing reference picture, default is 28
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] Missing reference picture, default is 28
[h264 @ 0x7fb08c003100] reference picture missing during reorder
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] reference picture missing during reorder
[h264 @ 0x7fb08c003100] Missing reference picture, default is 30
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] Missing reference picture, default is 30
[h264 @ 0x7fb08c003100] illegal short term buffer state detected
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] illegal short term buffer state detected
[h264 @ 0x7fb08c003100] reference picture missing during reorder
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] reference picture missing during reorder
[h264 @ 0x7fb08c003100] Missing reference picture, default is 30
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] Missing reference picture, default is 30
[h264 @ 0x7fb08c003100] reference picture missing during reorder
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] reference picture missing during reorder
[h264 @ 0x7fb08c003100] Missing reference picture, default is 32
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] Missing reference picture, default is 32
[h264 @ 0x7fb08c003100] illegal short term buffer state detected
2019-05-06 05:31:23 ERROR  : [h264 @ 0x7fb08c003100] illegal short term buffer state detected
```

**Comment by nschlia** [comment]
I've seen that, too. Seems to be caused if you try to access a non-i-frame directly. I'll check how to fix that one. I tested with Prores i-frame only files and never had that problem, only with H264 mp4.

**Comment by zhuker** [comment]
it would be good to return some kind of error to `cp` if frame can not be decoded for any reason
(e.g. `ENOTSUP`?)

**Comment by nschlia** [comment]
> it would be good to return some kind of error to `cp` if frame can not be decoded for any reason
> (e.g. `ENOTSUP`?)

I am doing that actually, but it seems the error code gets lost inside ffmpegfs on the way. The messages come from FFmpeg, either from the decoder or the encoder. Seems it gets stuck on something. I'll figure that out (probably the decoder needs the i-frame to properly decode the p/b-frame.).

**Comment by nschlia** [comment]
It's like I have expected, for videos with p/b-frames I have to go back to the last i-frame and start decoding from that key frame. In you video, number 1, 27 and 54 are key frames. So if you want to open e.g. frame 37, the decoder needs to rewind to frame 27 and start decoding from there.

That's all. I'll implement that.

**Comment by nschlia** [comment]
* Fixed a bug that cause crashes due to GPFs/Bus Errors
* Fixed hangs because frames were recoded over an over

**Comment by nschlia** [comment]
Seek requests now go to the correct decoder thread.

**Comment by zhuker** [comment]
IT WORKS!!! 🥇 
magic 
thank you so much

if only i could copy files without listing directory first it would be the answer to our prayers

**Comment by nschlia** [comment]
```
:~/test/test$ cp -v /home/norbert/test/out/orig.mp4.mp4/00000000008.png .
'/home/norbert/test/out/orig.mp4.mp4/00000000008.png' -> './00000000008.png'
:~/test/test$ cp -v /home/norbert/test/out/orig.mp4.mp4/00000000018.png .
'/home/norbert/test/out/orig.mp4.mp4/00000000018.png' -> './00000000018.png'
:~/test/test$ cp -v /home/norbert/test/out/orig.mp4.mp4/00000000058.png .
'/home/norbert/test/out/orig.mp4.mp4/00000000058.png' -> './00000000058.png'
:~/test/test$ ll /home/norbert/test/out/orig.mp4.mp4/00000000008.png    
-rw-rw-rw- 1 norbert norbert 2715307 Mai  5 12:18 /home/norbert/test/out/orig.mp4.mp4/00000000008.png
:~/test/test$ ll /home/norbert/test/out/orig.mp4.mp4/00000000028.png 
-rw-rw-rw- 1 norbert norbert 40960 Mai  5 12:18 /home/norbert/test/out/orig.mp4.mp4/00000000028.png

```

Workz :)

**Comment by zhuker** [comment]
nice!

**Comment by nschlia** [comment]
Great. Think we can close this issue, has grown large enough... Any bugs or additional requests should be handled by new issues.
